src/streams/EditStreamCard.js

```diff
@@ -1,50 +1,257 @@
+            // switch back to the first setting without making them submit
+            // the accidental setting or restart the app.
+            disabled: x.key !== initialValue && disabledIfNotInitialValue,
+          };
+        }),
+    [
+      initialValue,
+      webPublicStreamsEnabled,
+      canCreateWebPublicStreams,
+      createWebPublicStreamPolicy,
+      enableSpectatorAccess,
+      realmName,
+      realmUrl,
+      role,
+    ],
+  );
+}
+
 export default function EditStreamCard(props: Props): Node {
-  const { navigation, onComplete, initialValues, isNewStream } = props;
+  const { navigation, initialValues, isNewStream } = props;
+  const _ = useContext(TranslationContext);
 
   const [name, setName] = useState<string>(props.initialValues.name);
   const [description, setDescription] = useState<string>(props.initialValues.description);
   const [privacy, setPrivacy] = useState<Privacy>(props.initialValues.privacy);
+  // When adding more, update areInputsTouched.
 
-  const handlePerformAction = useCallback(() => {
-    onComplete(name, description, privacy);
-  }, [onComplete, name, description, privacy]);
+  const [awaitingUserInput, setAwaitingUserInput] = useState<boolean>(true);
+  const areInputsTouched =
+    name !== initialValues.name
+    || description !== initialValues.description
+    || privacy !== initialValues.privacy;
 
-  const privacyOptions = useMemo(
-    () => [
-      { key: 'public', title: 'Public' },
-      { key: 'private', title: 'Private' },
-    ],
-    [],
+  useEffect(
+    () =>
+      /* $FlowFixMe[prop-missing]: Get Flow types for
+       react-navigation/react-navigation@6925e92dc */
+      navigation.addListener('beforeRemove', e => {
+        if (!(awaitingUserInput && areInputsTouched)) {
+          return;
+        }
+
+        e.preventDefault();
+
+        Alert.alert(
+          _('Discard changes?'),
+          _('You have unsaved changes. Leave without saving?'),
+          [
+            { text: _('Cancel'), style: 'cancel' },
+            {
+              text: _('Discard'),
+              style: 'destructive',
+
+              /* $FlowFixMe[prop-missing]: Get Flow types for
+               react-navigation/react-navigation@6925e92dc */
+              onPress: () => navigation.dispatch(e.data.action),
+            },
+          ],
+          { cancelable: true },
+        );
+      }),
+    [_, areInputsTouched, navigation, awaitingUserInput],
   );
 
+  const handlePerformAction = useCallback(async () => {
+    setAwaitingUserInput(false);
+    let result = false;
+    try {
+      if (props.isNewStream) {
+        result = await props.onComplete({ name, description, privacy });
+      } else {
+        result = await props.onComplete({
+          name: initialValues.name !== name ? name : undefined,
+          description: initialValues.description !== description ? description : undefined,
+          privacy: initialValues.privacy !== privacy ? privacy : undefined,
```

Comment:
stream settings: Improve how we decide if the user has changed a value
    
    Compare against the initial values from when the form was
    initialized, rather than against the current, live-updated stream,
    as we were doing before.

Does this logic actually work as stated? This `initialValues` comes from a prop, and on EditStreamScreen the prop comes from the stream's data in Redux. So it looks like this should also end up comparing to the current values in Redux, not the values the component started with.

I think something like this should do it (untested):
```diff
 export default function EditStreamCard(props: Props): Node {
-  const { navigation, initialValues, isNewStream } = props;
+  const { navigation, isNewStream } = props;
+  const initialValues = React.useRef(props.initialValues).current;
```

Reply:
How about using a ref in `EditStreamScreen` and keeping the `initialValues` name?

Reply:
> `initialValues` doesn't have the right name for what we're passing for it, I think: it shouldn't have that name if its contents can change to be different from the initial values.

Yeah, I agree.



> How about using a ref in `EditStreamScreen` and keeping the `initialValues` name?

Sure, that'd work.